<a href="https://www.cognitiveclass.ai"><img src = "https://cognitiveclass.ai/wp-content/themes/bdu3.0/static/images/cc-logo.png" align = left></a>

<br>
<br>

--------------------
# Search for Extra Terrestrial Intelligence (SETI)
###  SETI Signal Classification on PowerAI with Multi GPU
<hr>
<br>

### Introduction
In this Notebook, we will use the famous [SETI Dataset](https://github.com/setiQuest/ML4SETI/) to build a Convolutional Neural Networks capable to perform signals classification. CNN will say, with some associated error, what type of signal is the presented input. In this notebook, you will use IBM PowerAI with multiple GPU to train the model.

### Project overview:
Each night, using the Allen Telescope Array (ATA) in northern California, the SETI Institute scans the sky at various radio frequencies, observing star systems with known exoplanets, searching for faint but persistent signals. The current signal detection system is programmed to search only for particular kinds of signals: narrow-band carrier waves. However, the detection system sometimes triggers on signals that are not narrow-band signals  (with unknown efficiency) and are also not explicitly-known radio frequency interference (RFI). There seems to be various categories of these kinds of events that have been observed in the past.

The goal of SETI is to classify these signals accurately in real-time. This may allow the signal detection system to make better observational decisions, increase the efficiency of the nightly scans, and allow for explicit detection of these other signal types.

For more information refer to [SETI hackathon page](https://github.com/setiQuest/ML4SETI/).

### Formulating the problem as image classification:
Framing the radio signal data into spectrogram (a 2D visual representation), we can convert the problem into something akin to an image classification problem. Therefore we can run Convolutional Neural Network (CNN), will be run on the images which are the result of converting the signals to spectrogram.

### Training on Multi-GPU:
Running the CNN on large amound of data might take long time to train the model. Howerve, today, many systems contains Multiple GPUs for high performance computation. We can leverage this environments to run the training operation concurrently across multiple cards. One sample of these kind of environments is [IBM PowerAI](http://cocl.us/SETI-NIMBIX-PowerAI). In this notebook, we show you how to design and run your model on multiple GPUs.

In [1]:
import requests
import json
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import tensorflow as tf
import pickle
import time
#!sudo pip install sklearn
import os
from sklearn.metrics import confusion_matrix
from sklearn import metrics
from six.moves import urllib
import sys
import tarfile
import re
import math
import datetime

### Set the destination folder

In [2]:
### SET YOUR WORKING SPACE HERE! Use this folder to save intermediate results.
dataset_name = 'SETI_ds_64x128'
data_dir = "tmp/SETI1_data/"
train_dir = 'tmp/SETI1_train/'

# check point directory
chk_directory = train_dir + '/save2/'
checkpoint_path = chk_directory + 'model.ckpt'

# Create the  training and testing directories if not exist
if os.path.exists(data_dir) is False:
    os.makedirs(data_dir)
print data_dir


if os.path.exists(train_dir) is False:
    os.makedirs(train_dir)
print train_dir

tmp/SETI1_data/
tmp/SETI1_train/


### Import dataset reader
The signals for this notebook, have been converted to spectogram images, and stored as 4 files.
The following cell will load a python code that help us to decode the binary file, and read the SETI dataset.

In [3]:
!wget -q --output-document  SETI.zip  https://ibm.box.com/shared/static/jhqdhcblhua5dx2t7ixwm88okitjrl6l.zip
!unzip -o SETI.zip
import SETI

Archive:  SETI.zip
  inflating: SETI.py                 
  inflating: __MACOSX/._SETI.py      


### Download data
The dataset exist and shared on IBM box. Running the following cell, you can download the dataset and extract it.

In [4]:
def maybe_download_and_extract():
    DATA_URL =  'https://ibm.box.com/shared/static/qz33lcio9ip2j8qi2atxqs62gn3bnu2s.gz'
    dest_directory = data_dir
    if not os.path.exists(dest_directory):
        os.makedirs(dest_directory)
    filename = DATA_URL.split('/')[-1]
    filepath = os.path.join(dest_directory, filename)
    if not os.path.exists(filepath):
        #print 'No zip file exist in ', filepath
        def _progress(count, block_size, total_size):
            sys.stdout.write('\r>> Downloading %s %.1f%%' % (filename, float(count * block_size) / float(total_size) * 100.0))
        sys.stdout.flush()
        filepath, _ = urllib.request.urlretrieve(DATA_URL, filepath, _progress)

    statinfo = os.stat(filepath)
    print('Successfully downloaded', filename, statinfo.st_size, 'bytes.')
    extracted_dir_path = os.path.join(dest_directory, dataset_name)
    if not os.path.exists(extracted_dir_path):
        print 'Extracting to', extracted_dir_path
        tarfile.open(filepath, 'r:gz').extractall(dest_directory)
maybe_download_and_extract()

('Successfully downloaded', 'qz33lcio9ip2j8qi2atxqs62gn3bnu2s.gz', 22528713, 'bytes.')


### Load data SETI

In [5]:
ds_directory = data_dir + dataset_name
dataset = SETI.read_data_sets(ds_directory, one_hot=True, validation_size=0)
num_examples = dataset.train.num_examples
print 'Number of time series:', num_examples
dataset.train.images.shape

Extracting tmp/SETI1_data/SETI_ds_64x128/train-images-idx3-ubyte.gz
Extracting tmp/SETI1_data/SETI_ds_64x128/train-labels-idx1-ubyte.gz
Extracting tmp/SETI1_data/SETI_ds_64x128/test-images-idx3-ubyte.gz
Extracting tmp/SETI1_data/SETI_ds_64x128/test-labels-idx1-ubyte.gz
Number of time series: 3200


(3200, 8192)

### Understanding the imported data

The imported data can be divided as follow:

- Training (dataset.train) >>  Use the given dataset with inputs and related outputs for training of NN. In our case, if you give an image that you know that represents a "class1", this set will tell the neural network that we expect a "class1" as the output.  
        - 3,200 signals (images)
        - dataset.train.images for inputs
        - dataset.train.labels for outputs
  
  
- Test (dataset.test) >> the model does not have access to this informations prior to the test phase. It is used to evaluate the performance and accuracy of the model against "real life situations". No further optimization beyond this point.  
        - 8,00 signals
        - dataset.test.images for inputs
        - dataset.test.labels for outputs
        
        
#### Labels
- Each image (spectrum of signal) in the dataset has been labeled from 1 to 4, representing:
    - squiggle
    - narrowband
    - noise
    - narrowbanddrd

## Network Parameters

We place the parameters on CPU.   
__Notice:__ This code is not optimal for single-GPU training due to the cost of copying parameters between CPU and GPU.

In [6]:
# Parameters
decay_rate=0.92  #decay every 1500 steps with a base of 0.92:
decay_steps=1500
learning_rate = 0.005
training_epochs = 750
batch_size = 128
display_step = 100
moving_avg_decay = 0.9999     # The decay to use for the moving average.

#check point directory
chk_directory = train_dir +'/save/'
checkpoint_path = chk_directory + 'model.ckpt'


num_classes = 4 # number of possible classes for the problem


height = 64 # height of the image in pixels 
width = 128 # width of the image in pixels 
n_input = width * height # number of pixels in one image 
num_GPUs = 4

#log file directory
log_dir = train_dir + 'log-'+str(num_GPUs)+'-'+str(learning_rate)+'-'+str(decay_rate)+'-'+str(decay_steps)+'-0'

In following cell, we fill a queue with a portion of SETI images, to be used in each GPUs

In [7]:
def read_inputs():
    train_img_ds = dataset.train.images
    image = tf.convert_to_tensor(train_img_ds, np.float32)
    image = tf.reshape(image, [-1,height,width,1], name = 'images')
    train_lb_ds = dataset.train.labels
    label = tf.convert_to_tensor(train_lb_ds, np.float32)
    label = tf.reshape(label, [-1,4], name = 'labels')
    
    min_fraction_of_examples_in_queue = 0.2
    min_queue_examples = int(num_examples *  min_fraction_of_examples_in_queue)
    print ('Filling queue with %d SETI images before starting to train. '
    'This will take a few minutes.' % min_queue_examples)
    num_preprocess_threads = 16
    images, labels = tf.train.shuffle_batch([image, label], batch_size=batch_size,
        num_threads=num_preprocess_threads,
        capacity=min_queue_examples + 3 * batch_size,
        min_after_dequeue=min_queue_examples, enqueue_many=True)
    return(images,labels)

### How to use multi GPU for training?

PowerAI supports multi-GPU calculation.  Given multiple GPU cards, we run the SETI model on each GPU as following:

    1. Store all model parameters on the CPU, e.g weights and biases.
    2. Put a copy of the SETI model on each GPU, including conv2d, pre-activation, relu, etc.
    3. Divide up a large batch of data across the GPUs, and feed each model replica with a unique batch of data.
    4. Compute the inference on each GPU
    5. Calculate the gradients on each GPU
    6. Wait until all GPUs finish processing of their batches
    7. Update model parameters synchronously on CPU

### Variable definitions

We define all variables using __tf.get_variable()__ in order to place the variables on CPU to be shares across multiple GPU for training. 

In [8]:
def _variable_on_cpu(name, shape, initializer):
    with tf.device('/cpu:0'):
        var = tf.get_variable(name, shape, initializer=initializer, dtype=tf.float32)
    return var

In [9]:
def _variable_with_weight_decay(name, shape, stddev, wd):
    var = _variable_on_cpu( name, shape ,tf.truncated_normal_initializer(stddev=stddev, dtype=tf.float32))
    if wd is not None:
        weight_decay = tf.multiply(tf.nn.l2_loss(var), wd, name='weight_loss')
        tf.add_to_collection('losses', weight_decay)
    return var

### Inference

This function builds the graph as far as required for running the network forward to make predictions.

In [10]:
def inference(images):

    # conv1
    with tf.variable_scope('conv1') as scope: # a sample variable would be conv1/weights
        kernel = _variable_with_weight_decay('weights', shape=[5, 5, 1, 32], stddev=0.1,  wd=0.0)
        conv = tf.nn.conv2d(images, kernel, strides=[1, 1, 1, 1], padding='SAME')
        biases = _variable_on_cpu('biases', [32], tf.constant_initializer(0.1))
        pre_activation = tf.nn.bias_add(conv, biases)
        conv1 = tf.nn.relu(pre_activation, name=scope.name)

    # pool1
    pool1 = tf.nn.max_pool(conv1, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='SAME', name='pool1')
    # norm1
    norm1 = tf.nn.lrn(pool1, 4, bias=1.0, alpha=0.001 / 9.0, beta=0.75, name='norm1')

    # conv2
    with tf.variable_scope('conv2') as scope:
        kernel = _variable_with_weight_decay('weights', shape=[5, 5, 32, 64],  stddev=0.1,  wd=0.0)
        conv = tf.nn.conv2d(norm1, kernel, [1, 1, 1, 1], padding='SAME')
        biases = _variable_on_cpu('biases', [64], tf.constant_initializer(0.1))
        pre_activation = tf.nn.bias_add(conv, biases)
        conv2 = tf.nn.relu(pre_activation, name=scope.name)

    # norm2
    norm2 = tf.nn.lrn(conv2, 4, bias=1.0, alpha=0.001 / 9.0, beta=0.75,  name='norm2')
    # pool2
    pool2 = tf.nn.max_pool(norm2, ksize=[1, 2, 2, 1], strides=[1, 4, 4, 1], padding='SAME', name='pool2')
    
    
    # local3
    with tf.variable_scope('local3') as scope:
        # Move everything into depth so we can perform a single matrix multiply.
        reshape = tf.reshape(pool2, [batch_size, -1])
        dim = reshape.get_shape()[1].value
        weights = _variable_with_weight_decay('weights', shape=[dim, 1024], stddev=0.04, wd=0.004)
        biases = _variable_on_cpu('biases', [1024], tf.constant_initializer(0.1))
        local3 = tf.nn.relu(tf.matmul(reshape, weights) + biases, name=scope.name)

    # local4
    with tf.variable_scope('local4') as scope:
        weights = _variable_with_weight_decay('weights', shape=[1024, 256],  stddev=0.04, wd=0.004)
        biases = _variable_on_cpu('biases', [256], tf.constant_initializer(0.1))
        local4 = tf.nn.relu(tf.matmul(local3, weights) + biases, name=scope.name)
        #_activation_summary(local4)

    # linear layer(WX + b),
    # We don't apply softmax here because tf.nn.sparse_softmax_cross_entropy_with_logits accepts the unscaled logits and performs the softmax internally for efficiency.
    with tf.variable_scope('softmax_linear') as scope:
        weights = _variable_with_weight_decay('weights', [256, num_classes],  stddev=1/256.0, wd=0.0)
        biases = _variable_on_cpu('biases', [num_classes], tf.constant_initializer(0.0))
        softmax_linear = tf.add(tf.matmul(local4, weights), biases, name=scope.name)

    return softmax_linear

#### Loss function

As we want to run the CNN on multiple GPUs, we construct our model in a multi-tower fashion where each tower is assigned to a different GPU. The following function calculate the loss value for each tower.  


loss function, adds the ops required to generate loss, to the inference graph. It calculates the average cross entropy loss across the batch.  

__Notice:__ The total loss s defined as the cross entropy loss plus all of the weight decay terms (L2 loss).

In [11]:
# Calculate the total loss on a single tower running the SETI model.
def tower_loss(scope, logits, labels):
    # Build inference Graph.
    labels = tf.cast(labels, tf.int64)
    cross_entropy = tf.nn.softmax_cross_entropy_with_logits(labels=labels, logits=logits, name='cross_entropy_per_example')
    cross_entropy_mean = tf.reduce_mean(cross_entropy, name='cross_entropy')
    tf.add_to_collection('losses', cross_entropy_mean)

    # Assemble all of the losses for the current tower only.
    # 'losses' is the key for collection
    # scope is for e.g. 'tower_0'
    losses = tf.get_collection('losses', scope)

    # Calculate the total loss for the current tower.
    total_loss = tf.add_n(losses, name='total_loss')
    
    # Attach a scalar summary to all individual losses and the total loss; do the
    # same for the averaged version of the losses.
    for l in losses + [total_loss]:
        loss_name = re.sub('%s_[0-9]*/' % 'tower_GPU', '', l.op.name)
        tf.summary.scalar(loss_name, l)
    
    return total_loss

#### Averaging Gradients 

Calculating the gradients on each tower, results in 4 group of gradients, stored in __tower_grads__. The following  function takes __tower_grads__, and calculates the average gradient for each shared variable across all towers.
Notice: that this function provides a synchronization point across all towers.

 - tower_grads: List of lists of (gradient, variable) tuples. The outer list is over individual gradients. The inner list is over the gradient calculation for each tower.

This function returns list of pairs of (gradient, variable) where the gradient has been averaged across all towers.


In [12]:
def average_gradients(tower_grads):
    average_grads = []
    for grad_and_vars in zip(*tower_grads):
    # Note that each grad_and_vars looks like the following:
    #   ((grad0_gpu0, var0_gpu0), ... , (grad0_gpuN, var0_gpuN))
        grads = []
        for g, _ in grad_and_vars:
            # Add 0 dimension to the gradients to represent the tower.
            expanded_g = tf.expand_dims(g, 0)

            # Append on a 'tower' dimension which we will average over below.
            grads.append(expanded_g)

        # Average over the 'tower' dimension.
        grad = tf.concat(axis=0, values=grads)
        grad = tf.reduce_mean(grad, 0)

        # Keep in mind that the Variables are redundant because they are shared
        # across towers. So .. we will just return the first tower's pointer to
        # the Variable.
        v = grad_and_vars[0][1]
        grad_and_var = (grad, v)
        average_grads.append(grad_and_var)
    return average_grads

#### Training
 - The following cell, creates a graph to run one step of training with respect to the loss.
 - It creates an optimizer and apply to all trainable variables. Add moving
  average for all trainable variables.
 - When training a model, it is often recommended to lower the learning rate as the training progresses. 
This function applies an exponential decay function to a provided initial learning rate!

In [13]:
with tf.device('/cpu:0'):
    # These varibales are defined on CPU
    # with tf.Graph().as_default(), tf.device('/cpu:0'):
    # Create a variable to track the global step.
    global_step = tf.Variable(0, name='global_step', trainable=False)

    # create learning_decay
    lr = tf.train.exponential_decay( learning_rate,
                                     global_step,
                                     decay_steps,
                                     decay_rate, staircase=True )
    
    # Use the optimizer 
    optimizer = tf.train.GradientDescentOptimizer(lr)
    
    tf.summary.scalar('learning_rate', lr)


 - We choose a unique name for all operations within a tower. e.g. all operations in the first tower are prepended with tower_GPU0. We do it using name_scope(). It defines __operations__ with unique prefix name within each tower on the corresponding GPU. 
 - All variables are placed to the CPU and accessed via tf.get_variable (no variable on GPUs).
 - we create a **train_op**, where we apply:
     - Gradients on all weights and biases in CPU.
     - Moving Average on all trainable variables.
 - **train_op** would be called later in the training process.

In [14]:
# Get images and labels for the model.
images, labels = read_inputs()
batch_queue = tf.contrib.slim.prefetch_queue.prefetch_queue([images, labels], capacity=2 * num_GPUs)

# Calculate the gradients for each model tower.
with tf.device('/cpu:0'):
    towers_grads = []
    with tf.variable_scope(tf.get_variable_scope()): # It is defined to share/reuse all the variables on CPU.
        for i in xrange(num_GPUs):
            with tf.device('/gpu:%d' % i):
                with tf.name_scope('%s_%d' % ('tower_GPU', i)) as nscope:
                    # read one batch for the GPU
                    image_batch, label_batch = batch_queue.dequeue()
                    
                    # Build inference part of the Graph.
                    logits = inference(image_batch) # it builds a subgraph on twoer with all the operations, to use the shared variables.

                    # This function constructs the entire SETI model, calling inference.
                    # It shares the variables across all towers (e.g. conv1/weights).
                    # Calculate the loss for one tower of the SETI model. 
                    loss = tower_loss(nscope, logits, label_batch)
 
                    # For the next tower, when we call tower_loss() again, it calls inference(), and tries to create new variables.
                    # For example, it tries to create conv1/weights, again.
                    # We will reuse variables for the next tower. 
                    tf.get_variable_scope().reuse_variables()
                    
                    # Grads is a list over the gradient calculation for each tower.
                    # its shape looks like [(<tensor>,<variable>), (<tensor>,<variable>), ..]
                    # where, <tensor> is gradient value of <variable>
                    tower_grads = optimizer.compute_gradients(loss)
                    towers_grads.append(tower_grads)

    # calculate avg gradients on CPU                
    avg_grads = average_gradients(towers_grads)
    
    # Apply the gradients to adjust the shared variables
    apply_gradient_op = optimizer.apply_gradients(avg_grads, global_step=global_step)

    # Track the moving averages of all trainable variables
    with tf.name_scope('Moving_Avg_Decay'):
        variable_averages = tf.train.ExponentialMovingAverage(moving_avg_decay, global_step, name= 'variable_averages')
        variables_averages_op = variable_averages.apply(tf.trainable_variables())

    # Group all updates to into a single train op.
    # we call train_op in learning process
    # tf.group Creates an op that groups multiple operations.
    # When this op finishes, all ops in input have finished. This op has no output.
    train_op = tf.group(apply_gradient_op, variables_averages_op)

Filling queue with 640 SETI images before starting to train. This will take a few minutes.


You can check the constructed graph here:

In [15]:
# Helper functions for TF Graph visualization
# The following code is from Alex Mordvintsev deep dream notebook from
# https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/tutorials/deepdream/deepdream.ipynb


from IPython.display import clear_output, Image, display, HTML

def strip_consts(graph_def, max_const_size=32):
    """Strip large constant values from graph_def."""
    strip_def = tf.GraphDef()
    for n0 in graph_def.node:
        n = strip_def.node.add() 
        n.MergeFrom(n0)
        if n.op == 'Const':
            tensor = n.attr['value'].tensor
            size = len(tensor.tensor_content)
            if size > max_const_size:
                tensor.tensor_content = tf.compat.as_bytes("<stripped %d bytes>"%size)
    return strip_def
  
def rename_nodes(graph_def, rename_func):
    res_def = tf.GraphDef()
    for n0 in graph_def.node:
        n = res_def.node.add() 
        n.MergeFrom(n0)
        n.name = rename_func(n.name)
        for i, s in enumerate(n.input):
            n.input[i] = rename_func(s) if s[0]!='^' else '^'+rename_func(s[1:])
    return res_def
  
def show_graph(graph_def, max_const_size=32):
    """Visualize TensorFlow graph."""
    if hasattr(graph_def, 'as_graph_def'):
        graph_def = graph_def.as_graph_def()
    strip_def = strip_consts(graph_def, max_const_size=max_const_size)
    code = """
        <script>
          function load() {{
            document.getElementById("{id}").pbtxt = {data};
          }}
        </script>
        <link rel="import" href="https://tensorboard.appspot.com/tf-graph-basic.build.html" onload=load()>
        <div style="height:600px">
          <tf-graph-basic id="{id}"></tf-graph-basic>
        </div>
    """.format(data=repr(str(strip_def)), id='graph'+str(np.random.rand()))
  
    iframe = """
        <iframe seamless style="width:800px;height:620px;border:0" srcdoc="{}"></iframe>
    """.format(code.replace('"', '&quot;'))
    display(HTML(iframe))


graph_def = tf.get_default_graph().as_graph_def()
tmp_def = rename_nodes(graph_def, lambda s:"/".join(s.split('_',1)))
show_graph(tmp_def)

#### Evaluation
In the following cell we build all the operations for calculating the accuracy of the model

In [16]:
x_ph  = tf.placeholder(tf.float32, shape=[None, n_input], name = 'x_ph')
y_ph = tf.placeholder(tf.float32, shape=[None, num_classes], name = 'y_ph')
def build_evaluation_graph():
    with tf.variable_scope(tf.get_variable_scope()) as scope:
        tf.get_variable_scope().reuse_variables()
        image_test = tf.reshape(x_ph, [-1,height,width,1], name = 'image_test') 
        logits_test = inference(image_test)
        pred = tf.cast(logits_test, tf.float32)
        top_k_op = tf.nn.in_top_k(pred,tf.argmax(y_ph,1) , 1)
        return top_k_op
top_k_op = build_evaluation_graph()

In [17]:
def evaluate():
    with tf.variable_scope(tf.get_variable_scope()):
        num_iter = int(math.ceil(dataset.test.num_examples / batch_size))
        true_count = 0  # Counts the number of correct predictions.
        total_sample_count = num_iter * batch_size
        step = 0
        while step < num_iter:
            x_batch, y_batch = dataset.test.next_batch(batch_size)
            correct_count = np.sum(sess.run(top_k_op, feed_dict={x_ph: x_batch, y_ph: y_batch}))
            true_count += correct_count
            #print true_count
            step += 1

        # Compute precision @ 1.
        precision = true_count*1.0 / total_sample_count
        #tf.summary.scalar('precision', precision)
        return precision


### Create checkpoint directory

In [18]:
directory = os.path.dirname(chk_directory)
try:
    os.stat(directory)
    ckpt = tf.train.get_checkpoint_state(chk_directory)
    print ckpt
except:
    os.mkdir(directory) 

model_checkpoint_path: "tmp/SETI1_train//save/model.ckpt-18000"
all_model_checkpoint_paths: "tmp/SETI1_train//save/model.ckpt-400"
all_model_checkpoint_paths: "tmp/SETI1_train//save/model.ckpt-500"
all_model_checkpoint_paths: "tmp/SETI1_train//save/model.ckpt-600"
all_model_checkpoint_paths: "tmp/SETI1_train//save/model.ckpt-700"
all_model_checkpoint_paths: "tmp/SETI1_train//save/model.ckpt-18000"



## Training
Lets initialize all the variable, and resotre the previous trained model if have any

In [ ]:
# Initializing the variables
init = tf.global_variables_initializer()
loss_values = []

X_test = dataset.test.images
y_test = dataset.test.labels

sess = tf.Session(config=tf.ConfigProto(allow_soft_placement=True))
sess.run(init)
merged_op = tf.summary.merge_all()
train_writer = tf.summary.FileWriter(log_dir, sess.graph)
saver = tf.train.Saver(tf.global_variables())

# Start the queue runners.
tf.train.start_queue_runners(sess=sess)


# load previously trained model if appilcable
ckpt = tf.train.get_checkpoint_state(chk_directory)
if ckpt:
    print "loading model: ",ckpt.model_checkpoint_path
    #saver.restore(sess, ckpt.model_checkpoint_path)

loading model:  tmp/SETI1_train//save/model.ckpt-18000


In this step, we start training model by looping over all batches in each epoch, feeding the graph, and calculating the new weights.

In [ ]:
num_examples = dataset.train.num_examples
total_batch = int(num_examples / batch_size)
print 'Training dataset size:',num_examples, ' Signals'
print 'Testing dataset size:',dataset.test.num_examples, ' Signals'
print 'Total epochs:', training_epochs
print 'Total steps:', training_epochs * total_batch * num_GPUs
print 'Batch size:', batch_size
print 'Total batchs per epoch:',total_batch
print str(datetime.datetime.now())
# Training cycle
tr_start = time.time()
for epoch in range(training_epochs):
    ep_start = time.time()
    avg_accuracy = 0.

    # Loop over all batches in one epoch
    for step in range(total_batch):
        t_start = time.time()
        _, loss_value = sess.run([train_op, loss])
        duration = time.time() - t_start
        assert not np.isnan(loss_value), 'Model diverged with loss = NaN'
    
    
    # Save model every x epochs
    if epoch >= 0 and epoch % 100 == 0:
        saver.save(sess, checkpoint_path, global_step = epoch)
    
    # Summarize model every x epochs
    if epoch >= 0 and epoch % 10 == 0:
        #print 'merged'
        summary = sess.run(merged_op)    
        train_writer.add_summary(summary, epoch)
    
    
    # Display model every 1 epochs
    if epoch >= 0 and epoch % 50 == 0:
        plr = sess.run(lr)
        g_step = sess.run(global_step)
        loss_values.append(loss_value)
        sec_per_batch = duration / num_GPUs
        test_eval = evaluate()
        summ = tf.Summary()
        summ.ParseFromString(sess.run(merged_op))
        summ.value.add(tag='Precision @ 1', simple_value=test_eval)
        train_writer.add_summary(summ, epoch)
        print("Epoch:"+ '%04d' % (epoch) + \
        ", Ep_time=" + "{:.2f}".format(time.time() - ep_start) + \
        ", g_step:" + '%05d' % (g_step) + \
        ", lr=" + "{:.9f}".format(plr) + \
        ", cost=" + "{:.3f}".format(loss_value) + \
        #", Train_Acc=" + "{:.2f}".format(avg_train_acc) + \
        ", Test_Acc=" + "{:.2f}".format(test_eval) + 
        ", Batch (sec)=" + "{:.3f}".format(sec_per_batch)  ) 
        
print("Wall Time:","{:.1f}".format((time.time() - tr_start)/60.0, "Min"))
print("Optimization Finished!")
print ("model saved to {}".format(checkpoint_path))
saver.save(sess, checkpoint_path, global_step = (epoch+1)*step)
train_writer.close()

Training dataset size: 3200  Signals
Testing dataset size: 800  Signals
Total epochs: 750
Total steps: 75000
Batch size: 128
Total batchs per epoch: 25
2017-11-06 18:23:47.450077
Epoch:0000, Ep_time=30.84, g_step:00025, lr=0.005000000, cost=22.830, Test_Acc=0.25, Batch (sec)=0.067
Epoch:0050, Ep_time=7.40, g_step:01275, lr=0.005000000, cost=21.591, Test_Acc=0.26, Batch (sec)=0.066
Epoch:0100, Ep_time=21.19, g_step:02525, lr=0.004600000, cost=20.280, Test_Acc=0.47, Batch (sec)=0.068
Epoch:0150, Ep_time=7.33, g_step:03775, lr=0.004232000, cost=19.286, Test_Acc=0.62, Batch (sec)=0.066
Epoch:0200, Ep_time=29.66, g_step:05025, lr=0.003893440, cost=18.463, Test_Acc=0.68, Batch (sec)=0.063
Epoch:0250, Ep_time=7.20, g_step:06275, lr=0.003581965, cost=17.891, Test_Acc=0.70, Batch (sec)=0.066
Epoch:0300, Ep_time=30.04, g_step:07525, lr=0.003295408, cost=17.138, Test_Acc=0.67, Batch (sec)=0.062
Epoch:0350, Ep_time=7.22, g_step:08775, lr=0.003295408, cost=16.654, Test_Acc=0.68, Batch (sec)=0.066
E

Lets see how the coss value changes over time.

In [ ]:
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt
plt.plot([np.mean(loss_values[i:i+5]) for i in range(len(loss_values))])
plt.show()


## Evaluation
Accuracy is depend on the number of epoch that you set in partametrs part.

In [ ]:
with tf.variable_scope(tf.get_variable_scope()):
    tf.get_variable_scope().reuse_variables()
    num_iter = int(math.ceil(dataset.test.num_examples / batch_size))
    step = 0
    pred_dict=[]
    while step < num_iter:
        #print step
        x_batch, y_test = dataset.test.next_batch(batch_size)
        image_test = tf.reshape(x_batch, [-1,height,width,1], name = 'image_test') 
        logits_test = inference(image_test)
        pred = tf.cast(logits_test, tf.float32)
        pred_lbl = sess.run(pred)
        pred_dict.append((y_test, pred_lbl))
        step += 1
y_pred = [] # predicted
y_lb = [] # ground truth labels
y_pred_lb = [] # predicted labels
for item in pred_dict:
    for hotcode_val in item[0]:
        y_lb.append(np.argmax(hotcode_val)) # ground truth
    for hotcode_val in item[1]:
        y_pred.append(hotcode_val)
        y_pred_lb.append(np.argmax(hotcode_val)) # ground truth
print "Confusion Matrix:"
print metrics.confusion_matrix(y_true= y_lb, y_pred= y_pred_lb)
print "Precision:"
print metrics.classification_report(y_true= y_lb, y_pred= y_pred_lb)
print("Classification accuracy: %0.6f" % metrics.accuracy_score(y_true= y_lb, y_pred= y_pred_lb) )
print("Log Loss: %0.6f" % metrics.log_loss(y_true= y_lb, y_pred= y_pred))

### Benchmark:
- SETI_multi_gpu_train.py achieves ~72% accuracy after 75K steps.
- Speed: With batch_size 128.  
- __Notice:__ The model is not optimized to reach to highest accuracy, and you can achive better results tuning the parameters.

<table border="1" style="box-sizing: border-box; border-spacing: 30px; background-color: transparent; color: #333333; font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif; font-size: 12px;">
<tbody style="box-sizing: border-box;">
<tr style="box-sizing: border-box;">
<td style="box-sizing: border-box; padding: 3px; text-align:left;"><span style="box-sizing: border-box; font-weight: bold;">CPU Architecture</span></td>
<td style="box-sizing: border-box; padding: 3px; text-align:left;"><span style="box-sizing: border-box; font-weight: bold;">CPU cores&nbsp;</span></td>
<td style="box-sizing: border-box; padding: 3px; text-align:left;"><span style="box-sizing: border-box; font-weight: bold;">Memory&nbsp;</span></td>
<td style="box-sizing: border-box; padding: 3px; text-align:left;"><span style="box-sizing: border-box; font-weight: bold;">GPU&nbsp;</span></td>
<td style="box-sizing: border-box; padding: 3px; text-align:left;"><span style="box-sizing: border-box; font-weight: bold;">Step time (sec/batch)&nbsp;</span></td>
<td style="box-sizing: border-box; padding: 3px; text-align:left;"><span style="box-sizing: border-box; font-weight: bold;">&nbsp;Accuracy</span></td>
</tr>
<tr style="box-sizing: border-box;">
<td style="box-sizing: border-box; padding: 3px; text-align:left;">POWER8</td>
<td style="box-sizing: border-box; padding: 3px; text-align:center;">160</td>
<td style="box-sizing: border-box; padding: 3px; text-align:center;">1 TB</td>
<td style="box-sizing: border-box; padding: 3px; text-align:left;">4 x Tesla K80</td>
<td style="box-sizing: border-box; padding: 3px; text-align:center;">~0.066 </td>
<td style="box-sizing: border-box; padding: 3px; text-align:left;">~72% at 75K steps  (83 minutes)</td>
</tr>
<tr style="box-sizing: border-box;">
<td style="box-sizing: border-box; padding: 3px; text-align:left;" >POWER8</td>
<td style="box-sizing: border-box; padding: 3px; text-align:center;">64</td>
<td style="box-sizing: border-box; padding: 3px; text-align:center;">256 GB</td>
<td style="box-sizing: border-box; padding: 3px; text-align:left;">2 x Tesla P100 w/NVLink np8g4</td>
<td style="box-sizing: border-box; padding: 3px; text-align:center;">~0.033 </td>
<td style="box-sizing: border-box; padding: 3px; text-align:left;">~72% at 75K steps  (40 minutes) </td>
</tr>
<tr style="box-sizing: border-box;">
<td style="box-sizing: border-box; padding: 3px; text-align:left;">POWER8</td>
<td style="box-sizing: border-box; padding: 3px; text-align:center;">128</td>
<td style="box-sizing: border-box; padding: 3px; text-align:center;">512 GB</td>
<td style="box-sizing: border-box; padding: 3px; text-align:left;">4 x Tesla P100 w/NVLink np8g4</td>
<td style="box-sizing: border-box; padding: 3px; text-align:center;">~0.017 </td>
<td style="box-sizing: border-box; padding: 3px; text-align:left;">~72% at 75K steps  (20 minutes)</td>
</tr>
</tbody>
</table>

## Want to learn more?

[Deep Learning with TensorFlow](http://cocl.us/SETI-NIMBIX-ML0102EN) is a free course in __cognitiveclass.ia__ where you can learn TensorFlow and Deep Learning togetherwe.

Also, running deep learning programs usually needs a high performance platform. PowerAI speeds up deep learning and AI. Built on IBM's Power Systems, PowerAI is a scalable software platform that accelerates deep learning and AI with blazing performance for individual users or enterprises. The PowerAI platform supports popular machine learning libraries and dependencies including Tensorflow, Caffe, Torch, and Theano. You can download a [free version of PowerAI](http://cocl.us/SETI-NIMBIX-PowerAI).

### Authors

<div class="teacher-image" style="    float: left;
    width: 115px;
    height: 115px;
    margin-right: 10px;
    margin-bottom: 10px;
    border: 1px solid #CCC;
    padding: 3px;
    border-radius: 3px;
    text-align: center;"><img class="alignnone wp-image-2258 " src="https://ibm.box.com/shared/static/tyd41rlrnmfrrk78jx521eb73fljwvv0.jpg" alt="Saeed Aghabozorgi" width="178" height="178" /></div>
#### Saeed Aghabozorgi

[Saeed Aghabozorgi](https://ca.linkedin.com/in/saeedaghabozorgi), PhD is Sr. Data Scientist in IBM with a track record of developing enterprise level applications that substantially increases clients’ ability to turn data into actionable knowledge. He is a researcher in data mining field and expert in developing advanced analytic methods like machine learning and statistical modelling on large datasets.</p>

